src/bin/pgcopydb/ld_transform.c

```diff
@@ -385,10 +388,42 @@ stream_transform_file(char *jsonfilename, char *sqlfilename)
 
 	/*
 	 * We might have a last pending transaction with a COMMIT message to be
-	 * found in a a later file.
+	 * found in a a later file. In that case though, the last message read was
+	 * a WAL SWITCH message.
+	 *
+	 * It might happen that --endpos has been set to an LSN found in the middle
+	 * of a transaction, in that case we ignore the transaction and insert a
+	 * KEEPALIVE message with the LSN we have reached.
 	 */
-	if (currentTx->count > 0)
+	if (currentTx->count > 0 && metadata->action == STREAM_ACTION_SWITCH)
+	{
+		++txns.count;
+	}
+	else if (currentTx->count > 0 && metadata->action != STREAM_ACTION_COMMIT)
```

Comment:
_Minor comment_: `&& metadata->action != STREAM_ACTION_KEEPALIVE` as well, it's not a big deal as we will anyway regenerate `K` action if we enter into this condition.

Reply:
Good point, I'm leaning towards keeping the code that way for the moment though. As you're saying, adding a synthetic K message doesn't change much of anything, and the code is easier to read. Might revisit later though.